x-pack/plugins/ml/server/models/fields_service/fields_service.ts

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { APICaller } from 'kibana/server';
+
+/**
+ * Service for carrying out queries to obtain data
+ * specific to fields in Elasticsearch indices.
+ */
+export function fieldsServiceProvider(callAsCurrentUser: APICaller) {
```

Comment:
Building a multi-metric job against the gallery data set, using the full time range, and a small bucket span, I am seeing `circuit_breaking_exceptions` in the network tab of the dev console as I add extra influencers. 

```
{"statusCode":429,"error":"Too Many Requests","message":"[circuit_breaking_exception] [parent] Data too large, data for [<reused_arrays>] would be [988102840/942.3mb], which is larger than the limit of [986932838/941.2mb], real usage: [988102768/942.3mb], new bytes reserved: [72/72b], usages [request=147475312/140.6mb, fielddata=228074/222.7kb, accounting=1282468/1.2mb, inflight_requests=784/784b], with { bytes_wanted=988102840 & bytes_limit=986932838 & durability=\"TRANSIENT\" } (and) [circuit_breaking_exception]...
```

We need to try and optimize the queries being done in here. There are probably two factors that are making it use a lot of memory:

1. It’s doing all the fields in one query
2. It’s using the full time range

If the bucket span is small and the time range long, then that’s a lot of time buckets e.g. 15 minute bucket span, 12 month time range.

So we probably need to be pragmatic about capping the number of buckets that we take the max over.  Say last 1000 buckets. That should be easy to calculate knowing the bucket span and time range provided. Just adjust `earliestMs` to be `max(earliestMs, latestMs - 1000 * interval)` (obviously needs to be converted to ms).

As well as being defensive in the queries, you should probably also expose to the user that an error has occurred calling the estimate model memory limit endpoint e.g. in a toast notification. Currently you only see the error in the Kibana server console, or by looking in the browser network tab.

Reply:
Thanks for testing! For now, I added capping of the time range as well as replace `must` part of the query with `filter` to take advantage of caching, check https://github.com/elastic/kibana/pull/60376/commits/7ee91e796e317709a56508571d487b1a8a5d604d. 
In a follow-up PR, I'll try to split the current ES query into multiple (in case there are plenty of influencers provided) and check how it performs.